joblib/numpy_pickle.py

```diff
@@ -63,135 +47,428 @@ def _read_magic(file_handle):
+    # already written data to the file.
+    if mode == 'w+':
+        mode = 'r+'
 
+    marray = np.memmap(filename, dtype=dtype, shape=shape, order=order,
+        mode=mode, offset=offset)
+
+    return marray
+
+
+def _read_array(fp):
+    """
+    Read an array from an NPY file.
+
+    Parameters
+    ----------
+    fp : file_like object
+        If this is not a real file object, then this may take extra memory
+        and time.
+
+    Returns
+    -------
+    array : ndarray
+        The array from the data on disk.
+
+    Raises
+    ------
+    ValueError
+        If the data is invalid.
+
+    """
+    version = _read_numpy_magic(fp)
+    _check_version(version)
+    shape, fortran_order, dtype = _read_array_header(fp, version)
+    if len(shape) == 0:
+        count = 1
+    else:
+        count = np.multiply.reduce(shape)
+
+    # Now read the actual data.
+    if dtype.hasobject:
+        # The array contained Python objects. We need to unpickle the data.
+        array = pickle.load(fp)
+    else:
+        if isfileobj(fp):
+            # We can use the fast fromfile() function.
+            array = np.fromfile(fp, dtype=dtype, count=count)
+        else:
+            # This is not a real file. We have to read it the
+            # memory-intensive way.
+            # crc32 module fails on reads greater than 2 ** 32 bytes,
+            # breaking large reads from gzip streams. Chunk reads to
+            # BUFFER_SIZE bytes to avoid issue and reduce memory overhead
+            # of the read. In non-chunked case count < max_read_count, so
+            # only one read is performed.
+
+            max_read_count = BUFFER_SIZE // min(BUFFER_SIZE, dtype.itemsize)
+
+            array = np.empty(count, dtype=dtype)
+            for i in range(0, count, max_read_count):
+                read_count = min(max_read_count, count - i)
+                read_size = int(read_count * dtype.itemsize)
+                data = _read_bytes(fp, read_size, "array data")
+                array[i:i+read_count] = np.frombuffer(data, dtype=dtype,
+                                                      count=read_count)
+
+        if fortran_order:
+            array.shape = shape[::-1]
+            array = array.transpose()
+        else:
+            array.shape = shape
+
+    return array
 ###############################################################################
 # Utility objects for persistence.
 
-class NDArrayWrapper(object):
+
+class NPArrayWrapper(object):
     """ An object to be persisted instead of numpy arrays.
 
         The only thing this object does, is to carry the filename in which
         the array has been persisted, and the array subclass.
     """
-    def __init__(self, filename, subclass, allow_mmap=True):
-        "Store the useful information for later"
-        self.filename = filename
+
+    def __init__(self, subclass, allow_mmap=True, offset=c_int64(-1)):
```

Comment:
According to https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#class-docstring, `__init__` should be documented inside the class docstring.